Anima/diffusion/mcm_estimator/animaMCMEstimatorImageFilter.hxx

```diff
@@ -917,140 +957,185 @@ MCMEstimatorImageFilter<InputPixelType, OutputPixelType>
                 p[i] = workVec[i];
-            
-            double costValue = this->PerformSingleOptimization(p,cost,lowerBounds,upperBounds);
-            
-            if ((costValue < optimalCostValue)||(restartNum == 0))
+
+            double tmpValue = this->GetCostValue(cost,p);
+
+            if ((tmpValue < optValue) || (optValue < 0))
             {
-                // - Get estimated data
-                for (unsigned int i = 0;i < dimension;++i)
-                    workVec[i] = p[i];
-                
-                mcmOptimalModel->SetParametersFromVector(workVec);
-                
-                this->GetProfiledInformation(cost,mcmOptimalModel,optimalB0Value,optimalSigmaSqValue);
-                optimalCostValue = costValue;
+                optValue = tmpValue;
+                optKappa = tmpKappa;
+                optNu = tmpNu;
             }
         }
     }
+
+    for (unsigned int i = numIsoCompartments;i < numCompartments;++i)
+    {
+        mcmUpdateValue->GetCompartment(i)->SetOrientationConcentration(optKappa);
+        mcmUpdateValue->GetCompartment(i)->SetExtraAxonalFraction(optNu);
+    }
+}
+
+template <class InputPixelType, class OutputPixelType>
+void
+MCMEstimatorImageFilter<InputPixelType, OutputPixelType>
+::TensorCoarseGridInitialization(MCMPointer &mcmUpdateValue, CostFunctionBasePointer &cost,
+                                 MCMType::ListType &workVec,ParametersType &p)
+{
+    unsigned int dimension = p.GetSize();
+    unsigned int numIsoCompartments = mcmUpdateValue->GetNumberOfIsotropicCompartments();
+    unsigned int numCompartments = mcmUpdateValue->GetNumberOfCompartments();
+
+    double optWeight1 = 0.0;
+    double optWeight2 = 0.0;
+    double optAzimuth = 0.0;
+    double optValue = - 1.0;
+
+    double meanAxialDiff = 0.0;
+    double meanRadialDiff = 0.0;
+    unsigned int numUsefulCompartments = 0;
+
+    for (unsigned int i = numIsoCompartments;i < numCompartments;++i)
+    {
+        if (mcmUpdateValue->GetCompartmentWeight(i) > 0)
+        {
+            meanAxialDiff += mcmUpdateValue->GetCompartment(i)->GetAxialDiffusivity();
+            meanRadialDiff += mcmUpdateValue->GetCompartment(i)->GetRadialDiffusivity1();
+
+            ++numUsefulCompartments;
+        }
+    }
+
+    if (numUsefulCompartments > 0)
+    {
+        meanAxialDiff /= numUsefulCompartments;
+        meanRadialDiff /= numUsefulCompartments;
+    }
     else
     {
-        // For now essentially NODDI
-        
-        // Initialize from previous model, handled somewhere else
-        this->InitializeModelFromSimplifiedOne(mcmValue,mcmUpdateValue);
-        
-        double optNu = 0;
-        double optKappa = 0;
-        double optValue = 0;
-        
-        for (unsigned int j = 0;j < m_CoarseGridSize;++j)
+        meanAxialDiff = m_AxialDiffusivityFixedValue;
+        meanRadialDiff = (m_RadialDiffusivity1FixedValue + m_RadialDiffusivity2FixedValue) / 2.0;
+
+    }
+
+    meanAxialDiff -= 5.0e-4;
```

Comment:
I am ok with the tensor initialisation although I am not sure what are these `5.0e-4` everywhere. But I believe it is related to one of your comments earlier this week so I guess you know what you are doing here.